### ISSUE TITLE
Refactoring into kernels

### ISSUE THREAD

**Issue by mortendahl** [mortendahl/tf-encrypted#402]
Kernels (as current implemented in Pond and SecureNN):
- cache
- truncate
- reveal
- add
- reduce
- cumsum
- sub
- mul: (private, private), (private, public), (public, private), (public, public)
- square (think this could/should be removed)
- matmul
- conv2d
- avgpool2d
- batch_to_space_nd
- indexer
- transpose
- strided_slice
- split
- stack
- concat
- mask
- reshape
- expand
- squeeze
- equal
- zeros
- bitwise_not
- bitwise_and
- bitwise_or
- msb
- lsb
- bits
- negative
- non_negative
- less
- less_equal
- greater
- greater_equal
- select
- equal_zero
- relu
- maxpool2d
- maximum
- reduce_max
- argmax

**Comment by mortendahl** [comment]
One initial thought was to register (binary) kernels based on pattern `(op, type_left, type_right)` where `type_left` and `type_right` are among e.g. `int32` (public integer), `fixed16_replicated` (private fixedpoint), and `quant8_additive`(private quantised).

This fixes the precision, as eg 8 bit fixedpoint, and leaves implicit a few things (that could cause  assertions to fail):
- backing tensors used
- which players hold the shares (so far we've always assumed eg server0 and server1 but this should be generalised moving forward)

**Comment by mortendahl** [comment]
See also:
- https://github.com/mortendahl/tf-encrypted/issues/209
- https://github.com/mortendahl/tf-encrypted/issues/239
- https://github.com/mortendahl/tf-encrypted/issues/255
- https://github.com/mortendahl/tf-encrypted/issues/261

**Comment by ianlivingstone** [comment]
Where does this sit w.r.t roadmap? Is this a part of the tensor refactor or separate?

**Comment by jvmancuso** [comment]
@ianlivingstone this is the core of the tensor refactoring

**Comment by jvmancuso** [comment]
@mortendahl how would we handle n-ary operations?  assert same type for all args and handle the exception otherwise?

**Comment by mortendahl** [comment]
Yes I think we should do that for starters

**Comment by jvmancuso** [comment]
how does replicated SS from #401 fit into this? another set of types?

**Comment by mortendahl** [comment]
Some thoughts around how high level users can use different kernels.

Constraints:

- model specification should be free of cryptographic jargon

### Pond only

```python
model = tfe.keras.Sequential([
  tfe.keras.layers.Dense(activation='relu'),  # approximate when used with Pond
  tfe.keras.layers.Dense(activation='sigmoid'),  # approximate when used with Pond
])

model.compile(protocol='pond')
```

### Pond and SecureNN

```python
model = tfe.keras.Sequential([
  tfe.keras.layers.Dense(activation='relu'),  # will use Pond when compiled below
  tfe.keras.layers.Dense(activation='relu_exact'),  # will use SecureNN when compiled below
  tfe.keras.layers.Dense(activation='sigmoid'),  # will use Pond when compiled below
])

my_protocol = tfe.protocol.pond + {
  'relu_exact': tfe.protocol.securenn.relu
}

model.compile(protocol=my_protocol)
```

### Gazelle

```python
model = tfe.keras.Sequential([
  tfe.keras.layers.Dense(activation='relu'),
])

gazelle = {
  'dense': tfe.protocol.ahe.dense,
  'relu': tfe.protocol.aby.relu,
}

model.compile(protocol=gazelle)
```

**Comment by jvmancuso** [comment]
This looks really slick, and I think constructing dictionaries for protocols will be good practice.  A few thoughts:
- If we want to specify different protocols by layer, e.g. a Pond-approximated softmax for an attention layer but a SecureNN softmax for the output layer, how would we attempt that?
- Can we do this in such a way that paves the way for maintaining lower-level control over tensor ops? (e.g. not necessarily limiting to whatever's in `tfe.keras.layers`, to start)
- Can we still mix and match private/public ops here?

**Comment by mortendahl** [comment]
> If we want to specify different protocols by layer, e.g. a Pond-approximated softmax for an attention layer but a SecureNN softmax for the output layer, how would we attempt that?

Any thoughts on what the reason one might have to do this? Ie would the below be a reasonable framing of the motivation?

```python
model = tfe.keras.Sequential([
  tfe.keras.layers.Activation('softmax_approx'),
  tfe.keras.layers.Activation('softmax'),
])

my_protocol = {
  'softmax_approx': tfe.protocol.pond.softmax,
  'softmax': tfe.protocol.securenn.softmax,
}

model.compile(protocol=my_protocol)
model.fit(...)
```

> Can we do this in such a way that paves the way for maintaining lower-level control over tensor ops? (e.g. not necessarily limiting to whatever's in `tfe.keras.layers`, to start)

Yes I'm thinking that the above essentially gets translated down to something like:

```python
with tfe.protocol.pond.softmax:
  y = tfe.softmax(x)

with tfe.protocol.securenn.softmax:
  z = tfe.softmax(y)
```

meaning that `tfe.softmax` is essentially what is overloaded, and any code that sticks to that pattern (i.e. doesn't call kernels directly) will be automatically patched.

> Can we still mix and match private/public ops here? think I answered this one for myself, would be a matter of declaring that in the protocol definition I imagine

Regarding e.g. running a model on public weights and private inputs, then yes, this should work. `tfe.protocol.pond.matmul` will essentially have to be a dispatcher that figures out the exact kernel to invoke (out of the four possibilities in this case).

Regarding performing some layers in plaintext in e.g. a partial-secure computation, I'm wondering if we should add explicit conversion layers to account for this. They are already implicitly there in e.g. the Gazelle model.

**Comment by justin1121** [comment]
w.r.t your top level comment update:

> As a consequence of supporting operation overloading, tensors should be decoupled from any specific protocol and instead be simple data containers. One should still be able to perform, say `x + y`, and have it mapped to a current protocol for addition given the type of x and y.

How do you image this playing with dtypes? Perhaps we can just use the backing tensors dtypes (e.g. int64, int100, etc.) and then the protocol as the source of truth. Do you think we'd need a Replicated dtype for instance? Maybe would need a fixed point dtype to know whether to delegate to a truncated multiplication or not. I really like the idea of not having a bunch of different tensor classes and then just differentiating them via the dtype. Not sure how possible this is in practice though.

**Comment by jvmancuso** [comment]
>Any thoughts on what the reason one might have to do this? Ie would the below be a reasonable framing of the motivation?
A fast softmax with higher approximation error may be suitable in some places (e.g. early attention layers) but a slow one that's exact or has less error might be critical later on in the network.  Later layers' approximation error may blow up errors in the earlier layers due to the chain rule in backprop, similar to exploding gradients but with respect to gradient bias and not numerical stability

**Comment by mortendahl** [comment]
> > Any thoughts on what the reason one might have to do this? Ie would the below be a reasonable framing of the motivation?
>
> A fast softmax with higher approximation error may be suitable in some places (e.g. early attention layers) but a slow one that's exact or has less error might be critical later on in the network. Later layers' approximation error may blow up errors in the earlier layers due to the chain rule in backprop, similar to exploding gradients but with respect to gradient bias and not numerical stability

Gotcha. I think the above snippet captures it okay then, where the user is make the distinction explicit using an `approx` tag.

**Comment by mortendahl** [comment]
@justin1121 
> > As a consequence of supporting operation overloading, tensors should be decoupled from any specific protocol and instead be simple data containers. One should still be able to perform, say `x + y`, and have it mapped to a current protocol for addition given the type of x and y.
>
> How do you image this playing with dtypes? Perhaps we can just use the backing tensors dtypes (e.g. int64, int100, etc.) and then the protocol as the source of truth. Do you think we'd need a Replicated dtype for instance? Maybe would need a fixed point dtype to know whether to delegate to a truncated multiplication or not. I really like the idea of not having a bunch of different tensor classes and then just differentiating them via the dtype. Not sure how possible this is in practice though.

I think both makes sense, i.e. a `replicated` and `fixed`. My thinking around classes is that we should try to only have then when the physical representation is different:

- a `replicated` value consists of six tensors, which is different from an `additive` value with two tensors, and hence should be different classes
- a `fixed` value consists of a single (encoded) tensor, which is different from a quantised value which is represented by two values
- etc.

But also wondering if it could to have extended dtype living in `DataType x SharingType x BackingType`, where `DataType = { int32, fixed16, quant8 }`, `SharingType = { replicated, additive }`, and `BackingType = { ring64, ring100, odd }` etc. Note that sometimes we would want to ignore part of the extended type though, for instance a kernel for sigmoid representation which might access inputs of extended type `fixed16_any_any`.

**Comment by jvmancuso** [comment]
>I think the above snippet captures it okay then

Possibly, although an `_approx` tag may not be sufficient in the long run (e.g. what if people want to supply custom polynomial approximations)

>Yes I'm thinking that the above essentially gets translated down to something like: ...

I'd like to be able to do something like this as well:
```python
my_protocol = {
    "dense": tfe.protocol.pond.dense,
    "relu": tfe.protocol.securenn.relu,
    ...
}

with my_protocol:
    x = tfe.conv2d(x, 64, 3, 3, 1)
    x = tfe.avgpool2d(x, 2, 2)
    x = tfe.relu(x)
    x = tfe.reshape(x, [-1, 64 * 7 * 7])
    x = tfe.dense(x, 10)
```

This way, I'd be able to define my own layers with specific protocol requirements, e.g.
```python
def residual_block(x):
    with tfe.protocol.pond + {"relu": tfe.protocol.securenn.relu}:
        for i in range(2):
            y = tfe.conv2d(x, 64, 3, 3, 1)
            y = tfe.batchnorm(y)
            y = tfe.relu(y)
            y = tfe.dropout(y, 0.2)
        return y + x
```

We'd talked about keeping API at a high level, but this seems like an pretty intuitive way to be able to support both.

**Comment by justin1121** [comment]
@mortendahl 

> My thinking around classes is that we should try to only have then when the physical representation is different

Ahh yes I remember discussing this with you!

This all makes sense to me. Although ideally we could just have one I agree that just having one would be hiding too much information about the underlying representation. Extended dtype is cool and also your generic dtype `fixed16_any_any` is a neat idea as well.

So say you're in a Replicated Protocol:

```
with tfe.protocol.replicated:
  tfe.mul(x, y)
```

I think there are two routes we could go for the definition of `mul`:

1.
```
def mul(x, y):
  # construct protocol (e.g. replicated, pond) tensors from x and y
  # calls kernels["mul"][extended_dtypes]
```

2. 
```
def mul(x, y):
  # directly calls kernels["mul"][protocol](x, y)
  # then inside that call protocol tensors get created
```

I'm leaning towards number 1 because I think some of that code could be written pretty generic for every op. Perhaps there are more options I have no considered though.

**Comment by mortendahl** [comment]
@jvmancuso 

> > I think the above snippet captures it okay then
>
> Possibly, although an `_approx` tag may not be sufficient in the long run (e.g. what if people want to supply custom polynomial approximations)

in that case you could do something like

```python
model = tfe.keras.Sequential([
  tfe.keras.layers.Activation('softmax_approx'),
  tfe.keras.layers.Activation('softmax_approx_custom'),
  tfe.keras.layers.Activation('softmax'),
])

my_protocol = {
  'softmax_approx': tfe.protocol.pond.softmax,
  'softmax_approx_custom': tfe.protocol.pond.Softmax(coefs),
  'softmax': tfe.protocol.securenn.softmax,
}
```

> I'd like to be able to do something like this as well:
> ...

I like it!

**Comment by mortendahl** [comment]
@justin1121 

> So say you're in a Replicated Protocol:

What do you mean by a replicated protocol? I was thinking that "replicated" is just how data is represented and not a protocol in itself (maybe except for casting kernels for input/output).

Are you suggesting that `x` and `y` are of type `tfe.Tensor` with a dtype that gives the extended type as needed for kernel lookup?

In your snippet my thoughts would be that `x` and `y` are both, say, `Fixed` tensors, each with a reference to a `Replicated` tensor, which in turn hold references to six `Ring64` tensors. So `x.extended_dtype` is `(fixed16, replicated, ring64)`. Finding the right kernel is then a question of something like `kernels[op, x.extended_dtype, y.extended_dtype]`, potentially with some logic to match on prefixes of the types instead to account for kernels that don't care about low level types (say sigmoid approximation that only need to match on e.g. `fixed16`).

Another thought was that we could write the plumbing code only once by simply subclassing the tensor classes that we want to support kernels for, e.g.

```python
class Replicated(KernelSupported):
    ...
```

as opposed to tensor classes for which normal overloading is enough, e.g.

```python
class Ring():
    ...
```

**Comment by justin1121** [comment]
> What do you mean by a replicated protocol? I was thinking that "replicated" is just how data is represented and not a protocol in itself (maybe except for casting kernels for input/output)

It was mostly just an example protocol. But I was thinking replicated would be a protocol. How do you think we'd specify this type of data representation if its not a protocol?

> Are you suggesting that `x` and `y` are of type `tfe.Tensor` with a dtype that gives the extended type as needed for kernel lookup?

That's kind of what I was thinking. Now that I think about it more, what I was trying to get at was how do we "lift" a plaintext value into an encrypted protocol value. I think this would happen before any operations/kernels are run but some conversion operations would need to be inserted between ops that have different data representations/protocols.

> Another thought was that we could write the plumbing code only once by simply subclassing the tensor classes that we want to support kernels

This sounds pretty interesting. Not sure I can picture how this would work right now might need to think about this in depth some more.

One last thing: what are your thought about `Ring64` would this be separate or the same as the native tensors?

**Comment by justin1121** [comment]
One other thing: I was thinking the Tensor class composition would be `Replicated` -> `Fixed` -> `Ring`. I think I see why you want it `Fixed` -> `Replicated` -> `Ring` and I think I like it. Just to confirm. Instead of saying a `Replicated` number is made x number of `Fixed` numbers you want to say a `Fixed` number is represented by a `Replicated` number. So at the high level you wouldn't know about a `Replicated` number just a `Fixed`, `Int` or whatever else number.

**Comment by justin1121** [comment]
Also just realized that `aby` would be the protocol when you use the `Replicated` numbers. Perhaps there are other protocols that could share kernels with `aby`.

**Comment by mortendahl** [comment]
@justin1121 

> It was mostly just an example protocol. But I was thinking replicated would be a protocol. How do you think we'd specify this type of data representation if its not a protocol?
>
> That's kind of what I was thinking. Now that I think about it more, what I was trying to get at was how do we "lift" a plaintext value into an encrypted protocol value. I think this would happen before any operations/kernels are run but some conversion operations would need to be inserted between ops that have different data representations/protocols.

Input and conversion are still a bit up in the air. Maybe it makes sense to unify them under casting, with some mechanism for implicit casting when needed. Wouldn't mind playing a bit more with this first.

> One last thing: what are your thought about Ring64 would this be separate or the same as the native tensors?

I was thinking that this is simply what we call backing tensors, i.e. separate classes. I think the extra level is relevant since they take care of e.g. modular reductions, and there are still some optimisation techniques we haven't applied there (such as https://github.com/mortendahl/tf-encrypted/issues/310).

> One other thing: I was thinking the Tensor class composition would be Replicated -> Fixed -> Ring. I think I see why you want it Fixed -> Replicated -> Ring and I think I like it. Just to confirm. Instead of saying a Replicated number is made x number of Fixed numbers you want to say a Fixed number is represented by a Replicated number. So at the high level you wouldn't know about a Replicated number just a Fixed, Int or whatever else number.

Yes exactly. I'm speculating that most users would probably simply be interested in knowing that it's e.g. a private fixedpoint, and in most cases happy to ignore what particular form of "private" it is (e.g. replicated or additive). In some cases their might care though, and we certainly do wrt kernels, hence the notion of an extended dtype.

> Also just realized that aby would be the protocol when you use the Replicated numbers. Perhaps there are other protocols that could share kernels with aby.

Yes exactly. The recent paper with quantised values for instance also uses replicated sharing.

**Comment by justin1121** [comment]
> Input and conversion are still a bit up in the air. Maybe it makes sense to unify them under casting, with some mechanism for implicit casting when needed. Wouldn't mind playing a bit more with this first.

Wonder if we could something like:

```
with tfe.protocol.aby(x, y):
  tfe.mul(x, y)
```

Where the x and y are automatically converted if needed by `tfe.protocol.aby(x, y)`. Not immediately obvious to me if that would actually work in python but some implicit thing like that might be interesting.

> Yes exactly. I'm speculating that most users would probably simply be interested in knowing that it's e.g. a private fixedpoint

This might be a little crazy but it almost seems to be that it should just be called a private float sense that's what the fixedpoint representation is actually representing.
